Forge.TreeWalker/src/TreeSchemaValidator.cs

```diff
@@ -0,0 +1,275 @@
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemaString(string schema, string rules)
+        {
+            return await Task.Run(() => Validate(new List<object> { schema }, rules));
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemasString(string schemas, string rules)
+        {
+            return await Task.Run(() => Validate(new List<object> { schemas }, ForgeSchemaValidationRules));
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemaInPath(string path, string rules)
+        {
+            return await GetschemaFromPathAndValidate(path, rules);
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateMultipleSchemasInPath(string path, string rules)
+        {
+            return await GetAllSchemasInDirectoryAndValidate(path, rules);
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchema(ForgeTree schema, JSchema rules)
+        {
+            return await Task.Run(() => Validate(new List<object> { schema }, ForgeSchemaValidationRules));
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemas(IList<ForgeTree> schemas, JSchema rules)
+        {
+            return await Task.Run(() => Validate(ConvertSchemasToSchemaList(schemas), rules));
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemasString(string schemas, JSchema rules)
+        {
+            return await Task.Run(() => Validate(new List<object> { schemas }, ForgeSchemaValidationRules));
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemaInPath(string path, JSchema rules)
+        {
+            return await GetschemaFromPathAndValidate(path, rules);
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateMultipleSchemasInPath(string path, JSchema rules)
+        {
+            return await GetAllSchemasInDirectoryAndValidate(path, rules);
+        }
+
+        public async Task<Tuple<bool, IList<string>>> ValidateSchemaString(string schema, JSchema rules)
+        {
+            return await Task.Run(() => Validate(new List<object> { schema }, rules));
+        }
+
+        private static List<object> ConvertStringToSchemaList(string schemas, out bool success)
+        {
+            try
+            {
+                MemoryStream ms = new MemoryStream(Encoding.UTF8.GetBytes(schemas));
+                DataContractJsonSerializer ser = new DataContractJsonSerializer(typeof(List<ForgeTree>));
+                var schemaList = ser.ReadObject(ms) as List<object>;
+                ms.Close();
+                success = true;
+                return schemaList;
+            }
+            catch
+            {
+                success = false;
+                return new List<object>();
+            }
+        }
+        private Task<Tuple<bool, IList<string>>> Validate(List<object> schemas, object rules)
```

Comment:
>Task<Tuple<bool, IList<string>> [](http://example.com/codeflow?start=16&length=31)

Let's return a bool and give the list of errors as an out parameter.

Reply:
I'm not a huge fan of Tuples. Needing to create them on either side of method calls to use them. Getting the values from Item1 and Item2 feels weird and not type-safe.

I would be open to using Tuples if we were in C# 7.0, where you can simple say:

(bool, IList<string>) result = (true, errorList);

Forge supports both netstandard2.0 and .Net 4.6.2, since Anvil still uses 4.6.2. So we can't use the sleek new Tuple syntax.

Reply:
Yes you mentioned it in the email. I made it tuple since I thought it should be consisted with other implementations being async. Could change to out parameter now since we get rid of task and async method